File: swift_app/Sources/Features/Ear/EarSeedLoader.swift
import Foundation

public enum EarSeedLoaderError: Error, LocalizedError {
    case seedFileNotFound
    case invalidDocument

    public var errorDescription: String? {
        switch self {
        case .seedFileNotFound:
            return "ear_seed_v1.json not found"
        case .invalidDocument:
            return "ear seed banks missing or empty"
        }
    }
}

public actor EarSeedLoader {
    public static let shared = EarSeedLoader()
    private var cached: EarSeedDocument?

    public init() {}

    public func load() async throws -> EarSeedDocument {
        if let cached {
            return cached
        }
        let data = try loadRawData()
        let decoder = JSONDecoder()
        let doc = try decoder.decode(EarSeedDocument.self, from: data)
        if doc.bankA.isEmpty && doc.bankB.isEmpty {
            throw EarSeedLoaderError.invalidDocument
        }
        cached = doc
        return doc
    }

    public func clearCacheForTest() {
        cached = nil
    }

    private func loadRawData() throws -> Data {
        if let bundleURL = Bundle.module.url(forResource: "ear_seed_v1", withExtension: "json") {
            return try Data(contentsOf: bundleURL)
        }
        throw EarSeedLoaderError.seedFileNotFound
    }
}

File: flutter_app/lib/ear/ear_seed_loader.dart
import 'dart:convert';

import 'package:flutter/services.dart';

import 'ear_seed_models.dart';

/// 从 Asset 加载 `ear_seed_v1.json`（内存缓存）。
class EarSeedLoader {
  EarSeedLoader._();

  static final EarSeedLoader instance = EarSeedLoader._();

  static const _assetPath = 'assets/data/ear_seed_v1.json';

  EarSeedDocument? _cached;

  /// 副作用：读 [rootBundle]；成功则缓存。
  Future<EarSeedDocument> load() async {
    if (_cached != null) return _cached!;
    final text = await rootBundle.loadString(_assetPath);
    final decoded = jsonDecode(text);
    if (decoded is! Map<String, dynamic>) {
      throw const FormatException('ear seed root is not an object');
    }
    final doc = EarSeedDocument.tryParseMap(decoded);
    if (doc == null) {
      throw const FormatException('ear seed banks missing or empty');
    }
    _cached = doc;
    return doc;
  }

  /// 仅测试：注入假数据并跳过 Asset。
  void debugSetCache(EarSeedDocument doc) {
    _cached = doc;
  }

  void clearCacheForTest() {
    _cached = null;
  }
}
